doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,84 @@
+# Linearizability
+
+## Summary
+
+Guarantee linearizability in the SQL layer by ensuring that any observed transaction will also be observed by any future transaction.
+Implement this by using a lower-bound timestamp for each timeline.
+Any function that generates a timestamp will either update the lower bound (if it generated a higher timestamp) or use the lower bound.
+
+## Goals
+
+We want to provide linearizability as defined by [Jepsen] which says that
+
+> every operation appears to take place atomically, in some order, consistent with the real-time ordering of those operations
+
+This is limited to transactions (or single statements) in the SQL layer.
+Any statement can use `AS OF` to ignore linearizability, which is only provided when materialize chooses a timestamp.
+
+[Jepsen]: http://jepsen.io/consistency/models/linearizable
+
+## Non-Goals
+
+We are not yet concerned about sinks that have sent data to other systems.
+That is, there is no defined relationship between what our SQL layer will produce and what data has or has not yet arrived to any sink.
+
+Any use of `AS OF` is not guaranteed to return data at the same timestamp across restarts.
```

Comment:
I'm not sure I follow this paragraph. The first sentence reference a "same timestamp" and I may misunderstand "same as what". If I issue the same `AS OF` query I would like to see the same results, or be told that I cannot (because compaction). What uses of `AS OF` might return data at different timestamps?